README.md
# Knife Push

[![Gem Version](https://badge.fury.io/rb/knife-push.svg)](https://badge.fury.io/rb/knife-push) [![Build Status](https://travis-ci.org/chef/knife-push.svg?branch=master)](https://travis-ci.org/chef/knife-push)

**Umbrella Project: [Knife](https://github.com/chef/chef-oss-practices/blob/master/projects/knife.md)**

**Project State: [Deprecated](https://github.com/chef/chef-oss-practices/blob/master/repo-management/repo-states.md#deprecated)**

**Issues [Response Time Maximum](https://github.com/chef/chef-oss-practices/blob/master/repo-management/repo-states.md): None**

**Pull Request [Response Time Maximum](https://github.com/chef/chef-oss-practices/blob/master/repo-management/repo-states.md): None**

The knife push plugin is used by the Chef workstation to interact with the Push API to start jobs, view job status, view job lists, and view node status.

## Requirements

- Chef 15.0 higher
- Ruby 2.5 or higher

## Installation

This plugin ships built into Chef Workstation. Install Chef Workstation to use the plugin.

## Configuration

If push server is running on the same host as the Chef Server, then no reconfiguration is required on the Chef workstation.

## Subcommands

This plugin provides the following Knife subcommands. Specific command options can be found by invoking the subcommand with a `--help` flag.

### job list

The `job list` subcommand is used to view a list of Push jobs.

#### Syntax $ knife job list

### job start

The `job start` subcommand is used to start a Push job.

#### Syntax

```shell
knife job start (options) COMMAND [NODE, NODE, ...]
```

#### Options

This argument has the following options:

`--timeout TIMEOUT`

The maximum amount of time (in seconds) by which a job must complete, before it will be stopped.

`-q QUORUM --quorum QUORUM`

The minimum number of nodes that match the search criteria, are available, and acknowledge the job request. This can be expressed as a percentage (e.g. 50%) or as an absolute number of nodes (e.g. 145). Default value: 100%

`-b --nowait`

Exit immediately after starting a job instead of waiting for it to complete.

`--with-env ENVIRONMENT`

Accept a json blob of environment variables and use those to set the variables for the client. For example '{"test": "foo"}' will set the push client environment variable "test" to "foo". (Push 2.0 and later)

`--in-dir DIR`

Execute the remote command in the directory DIR. (Push 2.0 and later)

`--file DATAFILE`

Send the file to the client. (Push 2.0 and later)

`--capture`

Capture stdin and stdout for this job. (Push 2.0 and later)

#### Examples

For example, to search for nodes assigned the role "webapp", and where 90% of those nodes must be available, enter:

```shell
knife job start -quorum 90% 'chef-client' --search 'role:webapp'
```

To search for a specific set of nodes (named chico, harpo, groucho, gummo, zeppo), and where 90% of those nodes must be available, enter:

```shell
knife job start --quorum 90% 'chef-client' chico harpo groucho gummo zeppo
```

Use the `knife job start` subcommand to run a job with the following syntax:

```shell
knife job start job_name node_name
```

For example, to run a job named add-glasses against a node named "ricardosalazar", enter the following:

```shell
knife job start add-glasses 'ricardosalazar'
```

### job output

The `job output` command is used to view the output of Push jobs. (Push 2.0 and later). The output capture flag must have been set on job start; see the --capture option.

#### Syntax

```shell
knife job output JOBID
```

#### Examples

```shell
knife job output 26e98ba162fa7ba6fb2793125553c7ae test --channel stdout
```

#### Options

--channel [stderr|stdout]

The output channel to capture.

### job status

The `job status` command is used to view the status of Push jobs.

#### Syntax

```shell
knife job status JOBID
```

#### Examples

For example, to view the status of a job that has the identifier of "235", enter:

```shell
knife job status 235
```

### node status

The `node status` argument is used to identify nodes that Push may interact with.

#### Syntax

```shell
knife node status
```

## Contributing

For information on contributing to this project see <https://github.com/chef/chef/blob/master/CONTRIBUTING.md>

## License

**Author:** John Keiser([jkeiser@chef.io](mailto:jkeiser@chef.io))

**Copyright:** Copyright Chef Software, Inc.

**License:** Apache License, Version 2.0

```text
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
# knife-push Change Log

<!-- latest_release -->
<!-- latest_release -->

<!-- release_rollup -->
<!-- release_rollup -->

<!-- latest_stable_release -->
## [v2.0.1](https://github.com/chef/knife-push/tree/v2.0.1) (2020-08-21)

#### Merged Pull Requests
- Optimize our requires [#77](https://github.com/chef/knife-push/pull/77) ([tas50](https://github.com/tas50))
<!-- latest_stable_release -->

## [v2.0.0](https://github.com/chef/knife-push/tree/v2.0.0) (2020-06-02)

#### Merged Pull Requests
- Update deprecated Net::HTTPServerException with Net::HTTPClientException [#74](https://github.com/chef/knife-push/pull/74) ([vsingh-msys](https://github.com/vsingh-msys))
- Require Chef Infra 15+ / Ruby 2.5+ [#75](https://github.com/chef/knife-push/pull/75) ([tas50](https://github.com/tas50))
- Update the readme to reflect current requirements [#76](https://github.com/chef/knife-push/pull/76) ([tas50](https://github.com/tas50))

## [v1.1.5](https://github.com/chef/knife-push/tree/v1.1.5) (2019-12-30)

#### Merged Pull Requests
- Convert PR testing to buildkite [#72](https://github.com/chef/knife-push/pull/72) ([tas50](https://github.com/tas50))
- Substitute require for require_relative [#71](https://github.com/chef/knife-push/pull/71) ([tas50](https://github.com/tas50))
- Support Ruby 2.4 in test [#73](https://github.com/chef/knife-push/pull/73) ([tas50](https://github.com/tas50))

## [v1.1.2](https://github.com/chef/knife-push/tree/v1.1.2) (2018-12-04)

#### Merged Pull Requests
- Don&#39;t ship the readme in the gem [#67](https://github.com/chef/knife-push/pull/67) ([tas50](https://github.com/tas50))

## [v1.1.1](https://github.com/chef/knife-push/tree/v1.1.1) (2018-12-03)

#### Merged Pull Requests
- Wire up Expeditor [#63](https://github.com/chef/knife-push/pull/63) ([tas50](https://github.com/tas50))
- Resolve all chefstyle warnings and test Chefstyle in Travis [#64](https://github.com/chef/knife-push/pull/64) ([tas50](https://github.com/tas50))
- Require Chef 13 or later [#65](https://github.com/chef/knife-push/pull/65) ([tas50](https://github.com/tas50))
- Slim the number of files we ship in the gem package [#66](https://github.com/chef/knife-push/pull/66) ([tas50](https://github.com/tas50))

lib/chef/knife/job_start.rb
# @copyright Copyright 2014 Chef Software, Inc. All Rights Reserved.
#
# This file is provided to you under the Apache License,
# Version 2.0 (the "License"); you may not use this file
# except in compliance with the License. You may obtain
# a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied. See the License for the
# specific language governing permissions and limitations
# under the License.
#

require_relative "job_helpers"

class Chef
  class Knife
    class JobStart < Chef::Knife

      include JobHelpers

      deps do
        require "chef/node"
        require "chef/search/query"
      end

      banner "knife job start <command> [<node> <node> ...]"

      option :run_timeout,
        long: "--timeout TIMEOUT",
        description: "Maximum time the job will be allowed to run (in seconds)."

      option :quorum,
        short: "-q QUORUM",
        long: "--quorum QUORUM",
        default: "100%",
        description: "Pushy job quorum. Percentage (-q 50%) or Count (-q 145)."

      option :search,
        short: "-s QUERY",
        long: "--search QUERY",
        required: false,
        description: "Solr query for list of job candidates."

      option :send_file,
        long: "--file FILE",
        default: nil,
        description: "File to send to job."

      option :capture_output,
        long: "--capture",
        boolean: true,
        default: false,
        description: "Capture job output."

      option :with_env,
        long: "--with-env ENV",
        default: nil,
        description: "JSON blob of environment variables to set."

      option :as_user,
        long: "--as-user USER",
        default: nil,
        description: "User id to run as."

      option :in_dir,
        long: "--in-dir DIR",
        default: nil,
        description: "Directory to execute the command in."

      option :nowait,
        long: "--nowait",
        short: "-b",
        boolean: true,
        default: false,
        description: "Rather than waiting for each job to complete, exit immediately after starting the job."

      option :poll_interval,
        long: "--poll-interval RATE",
        default: 1.0,
        description: "Repeat interval for job status update (in seconds)."

      def run
        job_name = @name_args[0]
        if job_name.nil?
          ui.error "No job specified."
          show_usage
          exit 1
        end

        @node_names = process_search(config[:search], name_args[1, @name_args.length - 1])

        job_json = {
          "command" => job_name,
          "nodes" => @node_names,
        }

        job_json["quorum"] = get_quorum(config[:quorum], @node_names.length)

        v2_json = job_json.dup

        v2_json["file"] = "raw:" + file_helper(config[:send_file]) if config[:send_file]

        v2_json["capture_output"] = config[:capture_output]
        env = get_env(config)
        v2_json["env"] = env if env
        v2_json["dir"] = config[:in_dir] if config[:in_dir]
        v2_json["user"] = config[:as_user] if config[:as_user]

        begin
          job = run_helper(config, v2_json)
        rescue Net::HTTPClientException => e
          raise e if e.response.code != "400"

          ui.warn "Falling back to Push Jobs v1 mode."

          unless env.empty?
            ui.error "Sending Environment attributes isn't possible for Push Jobs 1.0"
            exit 1
          end

          %i{ send_file dir user }.each do |feature|
            if config[feature]
              ui.error "Can't use a 2.0 feature (#{feature}) with a 1.0 server"
              exit 1
            end
          end

          job = run_helper(config, job_json)
        end

        output(job)

        exit(status_code(job))
      end

    end
  end
end
